File: medium/2641_cousins_in_binary_tree_ii/solution.cpp
#include "../../utils/cpp/nodes.hpp"
#include <vector>

class Solution {
public:
  utils::TreeNode *replaceValueInTree(utils::TreeNode *root) {
    root->val = 0;

    std::vector<utils::TreeNode *> level = {root};
    while (!level.empty()) {
      std::vector<utils::TreeNode *> nextLevel;

      for (const utils::TreeNode *node : level) {
        if (node->left)
          nextLevel.push_back(node->left);
        if (node->right)
          nextLevel.push_back(node->right);
      }

      int levelSum = 0;
      for (const utils::TreeNode *node : nextLevel)
        levelSum += node->val;

      for (const utils::TreeNode *node : level) {
        int levelDeduction = 0;

        if (node->left)
          levelDeduction += node->left->val;
        if (node->right)
          levelDeduction += node->right->val;

        if (node->left)
          node->left->val = levelSum - levelDeduction;
        if (node->right)
          node->right->val = levelSum - levelDeduction;
        ;
      }

      level = nextLevel;
    }

    return root;
  }
};

File: medium/2641_cousins_in_binary_tree_ii/solution.py
from typing import Optional

from utils.python.nodes import TreeNode


class Solution:
    def replaceValueInTree(
        self, root: Optional[TreeNode]
    ) -> Optional[TreeNode]:
        stack = [(root, None, 0)]

        while stack:
            level = stack
            stack = []

            level_sum = 0

            for i in range(len(level)):
                level_sum += level[i][0].val

            for i in range(len(level)):
                level[i][0].val = level_sum - level[i][0].val - level[i][2]

            while level:
                node, _, _ = level.pop()

                if node.left:
                    stack.append(
                        (node.left, node, node.right.val if node.right else 0)
                    )
                if node.right:
                    stack.append(
                        (node.right, node, node.left.val if node.left else 0)
                    )

        return root
